apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex

```diff
@@ -23,60 +25,215 @@ defmodule ArgosAggregation.Thesauri do
+
+    def request_single_node(id) do
+      ElasticSearch.DataProvider.get_doc("concept_#{id}")
+    end
+
+  end
+
+
+  defmodule DataProvider do
     require Logger
 
+    alias ArgosAggregation.ElasticSearch.Indexer
+    alias ArgosAggregation.Thesauri.ConceptParser
+
+    def get_by_date(%Date{} = date) do
+      date
+      |> stream_pages
+      |> Stream.map(fn element ->
+        case element do
+          {:error, _err} -> element
+          _ -> ConceptParser.Search.parse_single_doc(element)
+        end
+      end)
+    end
+
+    defp stream_pages(date) do
+      Stream.resource(
+        fn ->
+          case DataSourceClient.Http.request_by_date(date) do
+            {:ok, xml} -> ConceptParser.Search.load_first_page_url(xml)
+            error -> {[error], nil}
+          end
+        end,
+        fn page_url ->
+          case page_url do
+            nil -> {:halt, page_url}
+            page_url -> load_next_page(page_url)
+          end
+        end,
+        fn _val ->
+          Logger.info("Done reading thesaurus updates")
+        end
+      )
+    end
+
+    defp load_next_page(page_url) do
+      case DataSourceClient.Http.read_from_url(page_url) do
+        {:ok, xml} -> ConceptParser.Search.load_next_page_items(xml)
+        error -> {[error], nil}
+      end
+    end
+
+    def get_all() do
+      with {:ok, xml} <- DataSourceClient.Http.request_root_level() do
+        stream_read_hierarchy(xml)
+        |> Stream.flat_map(fn elements ->
+          case elements do
+            {:error, _} -> [elements]
+            _ -> ConceptParser.Hierarchy.read_list_of_descriptions(elements)
+          end
+        end)
+      end
+    end
+
+    defp stream_read_hierarchy(xml) do
+      Stream.resource(
+        fn ->
+          Logger.info("Start reading root level")
+          ConceptParser.Hierarchy.read_root_level(xml)
+        end,
+        #next_fun
+        fn (root_data) ->
+          case root_data do
+            {:error, error} ->
+              Logger.error(error)
+              {[{:error, error}], nil}
+            # last case roots are empty
+            nil ->
+              Logger.info("stopped becaus of nil")
+              {:halt, root_data}
+            [] ->
+              Logger.info("stopped because of empty list")
+              {:halt, root_data}
+            # process roots until all are empty
+            roots ->
+              load_next_nodes(roots)
+          end
+        end,
+        #end_fun
+        fn (_root_data) ->
+          Logger.debug("Finished tree traversing.")
+        end
+      )
+    end
+
+    defp load_next_nodes([ head | tail ]) do
+      # load complet hirarchy of next
+      Logger.info("Load next master branch #{head}")
+      with {:ok, id } <- ConceptParser.Utils.get_resource_id_from_uri(head),
+        {:ok, xml} <- DataSourceClient.Http.request_node_hierarchy(id),
+        {:ok, xml} <- ConceptParser.Utils.check_validity(xml) do
+          {[xml], tail}
+      end
```

Comment:
Du fängst weiter oben den nil case in der Stream.resource ab, weil hier kein else kommt. Wenn du das hier noch einfügen würdest, könntest du vielleicht noch mehr Feedback loggen? 

Aktuell wird quasi jeder Fehler in den 3 Funktionen hier "geschluckt" und einfach im weiteren Verlauf als nil weitergegeben.

Reply:
Hast recht, ist angepasst.